crates/subspace-farmer/src/plotting.rs

```diff
@@ -0,0 +1,323 @@
+                        let ArchivedSegment {
+                            root_block,
+                            mut pieces,
+                            object_mapping,
+                        } = archived_segment;
+                        let piece_index_offset = merkle_num_leaves * root_block.segment_index();
+
+                        let object_mapping =
+                            create_global_object_mapping(piece_index_offset, object_mapping);
+
+                        // TODO: Batch encoding
+                        for (position, piece) in pieces.iter_mut().enumerate() {
+                            if let Err(error) =
+                                subspace_solving.encode(piece_index_offset + position as u64, piece)
+                            {
+                                error!("Failed to encode a piece: error: {}", error);
+                                continue;
+                            }
+                        }
+
+                        if let Some(plot) = weak_plot.upgrade() {
+                            let pieces = Arc::new(pieces);
+                            // TODO: There is no internal mapping between pieces and their indexes yet
+                            // TODO: Then we might want to send indexes as a separate vector
+                            if let Err(error) = runtime_handle
+                                .block_on(plot.write_many(Arc::clone(&pieces), piece_index_offset))
+                            {
+                                error!("Failed to write encoded pieces: {}", error);
+                            }
+                            if let Err(error) = runtime_handle.block_on(
+                                commitments.create_for_pieces(&pieces, piece_index_offset),
+                            ) {
+                                error!("Failed to create commitments for pieces: {}", error);
+                            }
+                            if let Err(error) = object_mappings.store(&object_mapping) {
+                                error!("Failed to store object mappings for pieces: {}", error);
+                            }
+
+                            let segment_index = root_block.segment_index();
+                            last_root_block.replace(root_block);
+
+                            info!(
+                                "Archived segment {} at block {}",
+                                segment_index, block_to_archive
+                            );
+                        }
+                    }
+
+                    if let Some(last_root_block) = last_root_block {
+                        if let Some(plot) = weak_plot.upgrade() {
+                            if let Err(error) =
+                                runtime_handle.block_on(plot.set_last_root_block(&last_root_block))
+                            {
+                                error!("Failed to store last root block: {:?}", error);
```

Comment:
This was present before this branch. Should I remove this pretty print along with the ones that I added?

Reply:
If it compiles without it - yes